### ISSUE TITLE
Return the matching search value in a multi-string search (i.e. IndexOfAny with SearchValues<string>)

### ISSUE THREAD

**Issue by MihaZupan** [dotnet/runtime#110670]
### Discussed in https://github.com/dotnet/runtime/discussions/110663

<div type='discussions-op-text'>

<sup>Originally posted by **f2bo** December 12, 2024</sup>
.NET 9 introduced multi-string search with `IndexOfAny(this ReadOnlySpan<char> span, SearchValues<string> values)`, which is great if you need to find a set of matching strings. However, and I would say more often than not, in addition to the location, you also need to know which of the search values was found, for example, to count instances for each of the search values or even just to advance a pointer past the match and continue processing.

Until now, when searching for single characters, this was trivial since the matching search value had length 1 and could be directly retrieved from the location returned by `IndexOfAny`. When searching for strings, however, it now becomes necessary to execute an additional comparison at the match location to determine which of the search values was found. 

I'm aware that you could get this information by using a `Regex` instead, but it seems odd given that `IndexOfAny` has already determined the matching search value. Wouldn't it be possible to also return it, even if it's in the form of an index into the `ReadOnlySpan<string>` that was originally used to create the `SearchValues` instance. This could be provided either as an additional output parameter or by returning a tuple that also includes the location.

</div>

**Comment by MihaZupan** [comment]
This question came up a few times, including on the initial API proposal #85573 and during API reiview.

The reason why we chose not to expose the length/value that matched is that doing so means we have to choose a policy for which values we consider while matching.
E.g. for `"foobar".IndexOfAny(["foo", "foobar"])`, do we return the shortest vs longest match.
Regex was also the main expected use case for the API at the time, and it can't currently take advantage of the extra information.

**Comment by f2bo** [comment]
@MihaZupan Thanks for replying.

Regarding `index.IndexOfAny(["foo", "foobar"])`, my thinking is that if you include both in you search values, then it's because you consider them to be different and need to handle them separately, so it should return the longer match "foobar". If you wanted the other behavior, you would only include "foo" in your `SearchValues` and then disambiguate in your code.

**Comment by bencyoung-Fignum** [comment]
How about a seperate method to call with different approaches for longest/shortest once the index is returned?

**Comment by f2bo** [comment]
@bencyoung-Fignum 

> How about a seperate method to call with different approaches for longest/shortest once the index is returned?

I believe a separate method would have to involve a new comparison. It would have the starting location so it would be more efficient, and probably more optimized than the code one could quickly write, but it's not taking advantage of the fact that `IndexOfAny` already determined the matching value.

**Comment by dariusclay** [comment]
This would be extremely useful to know which value, or possible values, were matched.

**Comment by dariusclay** [comment]
Providing SearchValues.Create overload to indicate match behavior could also be considered, eg:

```cs
var searchValues = SearchValues.Create(values, stringComparison, matchBehavior);
```

**Comment by MihaZupan** [comment]
Thinking about this more, I'm not convinced making the policy configurable is worthwhile here. Matching the longest first value makes the most sense to me, and I also buy the argument https://github.com/dotnet/runtime/issues/110670#issuecomment-2540160099 that if you didn't want that, you can change the set of values you pass to `Create`.

That leaves us with the API shape. The first question is what do we return:
1. The matched `string` value
    - This opens the question of whether the returned value can be equivalent, or has to be the same. `SearchValues<string>` will currently normalize all values to their uppercase representation if you're using `OrdinalIgnoreCase`. The values would be equivalent under that comparison, but they may be different under `Ordinal`.
    - The nullability of the returned value is a bit unfortunate since we can't say "not null when the other return value is >= 0" if we exposed it as `int IndexOfAny(.., out string? matchedValue)` or `(int, string?) IndexOfAny(...)`.
2. The match length
    - This avoids the equivalent vs same issue
3. A custom `TValue` you'd pass to some `Create(KeyValuePair<string, TValue>[])` factory?
    - Sounds interesting, especially if you'd turn around and do a dictionary lookup on the matched value right away (e.g. if you were doing replacements), but possibly too niche

Once we decide on the returned value, there's the question of how to expose it:
1. Added `out` parameter on the same shape as before
    - ```c#
      int IndexOfAny(this ReadOnlySpan<char> span, SearchValues<string> values, out string? value);
      ```
2. Tuple return with a different method name since we can't overload on return type (feel free to suggest better names)
    - ```c#
      (int Index, string? Value) IndexOfAnyWithValue(this ReadOnlySpan<char> span, SearchValues<string> values);
      (int Index, string? Value) FindFirst(this ReadOnlySpan<char> span, SearchValues<string> values);
      ```
3. A `Try-` variant (feel free to suggest better names)
    - ```c#
      bool TryFindFirst(this ReadOnlySpan<char> span, SearchValues<string> values, out int index, out string? value);
      ```
    - This would solve the nullability issue since we could add `[NotNullWhen(true)]` on the `string?`
4. If we returned the match length, do we do it as two `int`s, or would `Range` be better

An implementation for the `int IndexOfAny(span, out string?)` variant is https://github.com/dotnet/runtime/compare/main...MihaZupan:searchvalues-matchedValue. The main change is due to the new policy, how/what we return is pretty simple to rewire.

Please share your thoughts and experience from potential usage places. Is having the value useful? Did you only ever need the length? ...

**Comment by dariusclay** [comment]
Agreed on longest-match-wins, and if you need that to change you use different values passed to Create.

Personally, a Range would give you exactly the information needed. With the start..end of the found value you can perform your own policy on top of it easily if needed (you know the set of values you passed to Create). Most of the time, it's just needing the span that was matched, so Range fits here as a return.

One naive way I was working around this was, once IndexOfAny returned a hit, I loop through the value set (ordered by longest) passed to Create and did a StartsWith comparison from the returned position to get which was found.

```cs
bool TryFindFirst(this ReadOnlySpan<char> span, SearchValues<string> values, out Range range)
```

Thoughts?

**Comment by MihaZupan** [comment]
> With the start..end of the found value you can perform your own policy on top of it easily if needed (you know the set of values you passed to Create).

Right, and that realistically involves a dictionary lookup. If that's a common need, it'd be an unfortunate (but not prohibitive) added cost given that you're using SearchValues because of perf in the first place.

**Comment by tannergooding** [comment]
@MihaZupan can you please update the top comment with a concrete proposed API, following the standard template?

This is "unactionable" otherwise and should be closed.

**Comment by MihaZupan** [comment]
We don't yet have a concrete proposal, it's not ready to move to api review - see gathering feedback on possible approaches outlined in https://github.com/dotnet/runtime/issues/110670#issuecomment-4692325032. I'm hoping for more potential users of such APIs to express their opinions.
